### ISSUE TITLE
consolidate java/scala content

### ISSUE THREAD

**Issue by patriknw** [akka/akka#23052]
With the `@java` and `@scala` tags and the multi-tab code snippets it's possible to consolidate java and scala content into same .md file and avoid duplication.

Example: https://raw.githubusercontent.com/akka/akka/master/akka-docs/src/main/paradox/scala/guide/tutorial_1.md

- [ ] actors.md - @richard-imaoka 
- [ ] ... *you can help too! Please write a comment what you're working on to avoid duplicate work.*

Can be done piece-by-piece.

**Comment by atiqsayyed** [comment]
@patriknw I'd be very happy to take this up, could you please explain a little more of what are we planning to do? And what are the areas that we'll be focussing on?

**Comment by raboof** [comment]
@atiqsayyed that would be great!

Roughly the process is as follows:
* Compare a file in `akka-docs/src/main/paradox/java/...` with the same one under `akka-docs/src/main/paradox/scala/...`. They should roughly cover the same contents, but one is for Java and the other is for Scala. The 'compare files' feature of IntelliJ IDEA is a nice tool for this.
* Update the file in the `scala` folder to contain both the Scala and Java content. You may use:
  * `@scala` and `@java` directives for inline text
  *  `@@@ div { .group-scala }` for blocks
  * [code blocks](http://developer.lightbend.com/docs/paradox/latest/features/groups.html) become tabbed
* Test your changes (`sbt akka-docs/paradocs`, see CONTRIBUTING.md for details)

You could look at https://github.com/akka/akka/pull/23039/files as an example - it's probably best to contribute PR's on a page-per-page basis at least at first. Feel free to let us know if you have any questions/problems!

**Comment by richard-imaoka** [comment]
Hi just to let you know and to avoid duplicated work, I already started on `paradox/scala/actors.md` and `paradox/java/actors.md`. I'm going to submit a PR soon. For other .md files I did not start the work yet.

**Comment by atiqsayyed** [comment]
Okay.. Unfortunately i also started with actors :D 
No Issues, I can work on other files. @richard-imaoka thanks for putting it here..

**Comment by ktoso** [comment]
Yep, thanks for coordinating :-)
I'll put a list who works on what in the first post. 
Are you going to pick up something else @atiqsayyed ?

**Comment by richard-imaoka** [comment]
@atiqsayyed I am really sorry about that ... I started working on actors.md about a week ago, so I should have submitted a PR earlier so that we could avoid the conflict.

Let's coordinate this using the below list (if the core team prefers, please move the list to the 1st comment on this issue). 

The core team or I can mark what @atiqsayyed is working on at the moment.

#### /java

- [ ] /java/actors.md - @richard-imaoka #23079 
- [ ] /java/agents.md
- [ ] /java/camel.md
- [ ] /java/cluster-client.md
- [ ] /java/cluster-metrics.md
- [ ] /java/cluster-sharding.md
- [ ] /java/cluster-singleton.md
- [ ] /java/cluster-usage.md
- [ ] /java/dev
- [ ] /java/dispatchers.md
- [ ] /java/distributed-data.md
- [ ] /java/distributed-pub-sub.md
- [ ] /java/event-bus.md
- [ ] /java/extending-akka.md
- [ ] /java/fault-tolerance-sample.md
- [ ] /java/fault-tolerance.md
- [ ] /java/fsm.md
- [ ] /java/futures.md
- [ ] /java/howto.md
- [ ] /java/index-actors.md
- [ ] /java/index-futures.md
- [ ] /java/index-network.md
- [ ] /java/index-utilities.md
- [ ] /java/index.md
- [ ] /java/io-tcp.md
- [ ] /java/io-udp.md
- [ ] /java/io.md
- [ ] /java/logging.md
- [ ] /java/mailboxes.md
- [ ] /java/persistence-query-leveldb.md
- [ ] /java/persistence-query.md
- [ ] /java/persistence-schema-evolution.md
- [ ] /java/persistence.md
- [ ] /java/remoting-artery.md
- [ ] /java/remoting.md
- [ ] /java/routing.md
- [ ] /java/scala-compat.md
- [ ] /java/scheduler.md
- [ ] /java/serialization.md
- [ ] /java/testing.md
- [ ] /java/typed-actors.md
- [ ] /java/typed.md

### java/stream

- [ ] /java/stream/index.md
- [ ] /java/stream/stages-overview.md
- [ ] /java/stream/stream-composition.md
- [ ] /java/stream/stream-cookbook.md
- [ ] /java/stream/stream-customize.md
- [ ] /java/stream/stream-dynamic.md
- [ ] /java/stream/stream-error.md
- [ ] /java/stream/stream-flows-and-basics.md
- [ ] /java/stream/stream-graphs.md
- [ ] /java/stream/stream-integrations.md
- [ ] /java/stream/stream-introduction.md
- [ ] /java/stream/stream-io.md
- [ ] /java/stream/stream-parallelism.md
- [ ] /java/stream/stream-quickstart.md
- [ ] /java/stream/stream-rate.md
- [ ] /java/stream/stream-testkit.md

**Comment by atiqsayyed** [comment]
I'll work on /java/agents.md
 /java/camel.md to start with and try to do as many as possible... 
If someone is picking anything please share it here.. Thanks!

**Comment by gosubpl** [comment]
For me:  `/java/distributed-data.md` :)

**Comment by gosubpl** [comment]
And I will also tackle two other: `/java/typed-actors.md` `/java/typed.md`

**Comment by richard-imaoka** [comment]
Sure, recorded @gosubpl's items, as well as @atiqsayyed's PR.

**Comment by atiqsayyed** [comment]
created pull request for stream-cookbook.md

**Comment by naferx** [comment]
hey @richard-imaoka , I'll give a try to `/java/stream/stream-integrations.md` and 
 `/java/stream/stream-introduction.md`

**Comment by richard-imaoka** [comment]
OK, marked the items for @atiqsayyed and @naferx

**Comment by atiqsayyed** [comment]
HI @richard-imaoka I'm picking these up next
-  /java/stream/stream-customize.md
-  /java/stream/stream-dynamic.md
-  /java/stream/stream-error.md

**Comment by richard-imaoka** [comment]
OK, marked the three with your name.

**Comment by naferx** [comment]
Hi again @richard-imaoka , I'm going to pick up the files below
  ``` 
 /java/stream/stream-graphs.md
 /java/stream/stream-io.md
 /java/stream/stream-parallelism.md
 /java/stream/stream-quickstart.md
 /java/stream/stream-rate.md
 /java/stream/stream-testkit.md
```
Thanks

**Comment by atiqsayyed** [comment]
Is there anything left @richard-imaoka ??

**Comment by richard-imaoka** [comment]
@atiqsayyed I updated the list above. Please pick ones no one is assigned yet.

**Comment by raboof** [comment]
Starting to looks great, only 8 more pages to go, and those have already been claimed by @richard-imaoka, @gosubpl, @atiqsayyed and @naferx! 

@richard-imaoka this seems to work really well, would you be interested in doing the same for akka-http? :)

**Comment by richard-imaoka** [comment]
Sure, will create an issue in akka-http to merge docs and follow up on it.

**Comment by atiqsayyed** [comment]
Sure why not 👍

**Comment by ktoso** [comment]
Thanks, that's awesome @richard-imaoka @atiqsayyed :)

**Comment by richard-imaoka** [comment]
Raised the above issue in akka-http 😃

**Comment by richardimaoka** [comment]
Now it looks like `/java/stream/index.md` is the last one 😃

**Comment by raboof** [comment]
Yes, excellent ;).

I couldn't resist trying out what the pages now look like when I remove the page refresh on language switch (https://github.com/akka/akka/commit/6dae33747c2909b18c243e54028744b9128401a6), sneak preview at http://pi.bzzt.net/akka-docs/ . Needs some more careful thinking/work before we can actually put it live, but you can experience how the language switching can now be nice and smooth there :)

**Comment by richardimaoka** [comment]
Yep, the new language switch is much smoother. I really like it! 👍 

As far as I can tell, remaining tasks are:

1. Redirect on the current versoin:  (e.g.) `/scala/actors.html` to `/actors.html`
2. Suggest redirect to the page without `/scala` or `/java` when you visit a page for an older version.

https://doc.akka.io/docs/akka/2.4/scala/actors.html
![2017-10-30_00h34_59](https://user-images.githubusercontent.com/7414320/32145351-54022d30-bd0a-11e7-9995-87a6188be8ba.png)

Is 1. more for server-side settings which the core team takes care, and 2. might be a trick in JavaScript the community can figure out?

**Comment by raboof** [comment]
In fact the redirection can be done in the repo, too - see https://github.com/akka/akka/pull/23544/files#diff-b5f788d092e9bf5e9eb7c8cb3365cf83 . I'll update that PR to resolve merge conflicts today and then perhaps we should just merge it.

I suppose we should make a separate ticket for re-introducing the 'suggest the latest version of this page' functionality? Though I think our SEO is now better so people don't end up on older versions of the docs so often anymore ;)

**Comment by raboof** [comment]
I added #23869 for the 'latest version' box. I think this one can now be closed! Many thanks to all of you who have been a tremendous help in achieving this!
